### PR TITLE
Feature/wmt 877/reductions worker

### PR DESCRIPTION
**Pull request by tsheppard01**
Adds reduction-worker.  Reduction worker task is called after create-workload task and in turn calls the calculate-workload-points task.

### CODE REVIEW

File: app/services/workers/reductions-worker.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        idsMap.set(status, ids)
+      }, this)
+
+      var updateReductionsPromises = []
+      for (var [status, ids] in idsMap) {
+        updateReductionsPromises.push(updateReductionStatusByIds(ids, status))
+      }
+      return Promise.all(updateReductionsPromises)
+    })
+}
```
Comment (ddebella) [inline app/services/workers/reductions-worker.js]: Should this not catch errors like we do the web?

File: app/services/data/get-all-open-reductions.js [hunk 2]
```diff
@@ -0,0 +1,9 @@
+const knexConfig = require('../../../knexfile').app
+const knex = require('knex')(knexConfig)
+
+module.exports = function () {
+  return knex('reductions')
+    .select()
+    .whereNull('status')
+    .orWhereIn('status', ['SCHEDULED', 'ACTIVE'])
```
Comment (ddebella) [inline app/services/data/get-all-open-reductions.js]: 👍

File: app/services/data/get-current-workload-report.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
+const config = require('../../../knexfile').app
+const knex = require('knex')(config)
+
+module.exports = function () {
+  return knex('workload_report')
+    .whereNull('effective_to')
```
Comment (caoimheb) [inline app/services/data/get-current-workload-report.js]: Also include where effective_from is not null? For consistency / extra check that it's a valid record

File: app/services/data/get-current-workload-report.js [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+    .select([
+      'id',
+      'effective_from'
+    ])
```
Comment (caoimheb) [inline app/services/data/get-current-workload-report.js]: Square brackets again?

File: app/services/data/get-current-workload-report.js [hunk 5]
```diff
@@ -0,0 +1,11 @@
+const config = require('../../../knexfile').app
```
Comment (caoimheb) [inline app/services/data/get-current-workload-report.js]: Slightly uncomfortable with the name of this file, as it's not really returning the current workload report. It's returning an array of open workload reports. Maybe get-open-workload-reports?

File: app/services/data/update-workload-report-with-status.js [hunk 6]
```diff
@@ -2,10 +2,10 @@ const knexConfig = require('../../../knexfile').app
...
 module.exports = function (workloadReportId, status) {
-  return knex('workload_report').where('id', workloadReportId)
-    .update({
-      effective_to: new Date(),
-      status: status,
-      status_description: status
-    })
+  return knex('workload_report')
+  .where('id', workloadReportId)
+  .update({
```
Comment (caoimheb) [inline app/services/data/update-workload-report-with-status.js]: 👍

File: app/services/workers/reductions-worker.js [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+        return createNewTasks([calculateWorkloadPointsTask])
+        .then(function () {
+          logger.info('CalculateWorkloadTask created')
+        })
+      })
+    })
+}
+
+var getReducitonStatus = function (reduction) {
```
Comment (caoimheb) [inline app/services/workers/reductions-worker.js]: Typo

File: app/services/workers/reductions-worker.js [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+  idsMap.set(reductionStatus.ACTIVE, [])
+  idsMap.set(reductionStatus.SCHEDULED, [])
+  idsMap.set(reductionStatus.DELETED, [])
+  idsMap.set(reductionStatus.ARCHIVED, [])
+
+  logger.info('Retrieving open reductions')
+  return getAllOpenReductions()
+    .then(function (reductions) {
+      reductions.forEach(function (reduction) {
+        status = getReducitonStatus(reduction)
```
Comment (caoimheb) [inline app/services/workers/reductions-worker.js]: Check does the status change before adding to the list of ids that need processed?

File: app/services/workers/create-workload.js [hunk 9]
```diff
@@ -85,18 +85,18 @@ module.exports.execute = function (task) {
...
                 workloadReportId,
                 undefined,
                 undefined,
                 taskStatus.PENDING
                 )
 
-      return createNewTasks([calculateWorkloadPointsTask])
+      return createNewTasks([reductionsWorkerTask])
             .then(function () {
               logger.info('Tasks created')
```
Comment (caoimheb) [inline app/services/workers/create-workload.js]: Add task type to logger info as in file below

File: test/integration/services/data/test-get-all-open-reductions.js [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+  })
+
+  it('should retrieve the open reductions in system', function (done) {
+    getAllOpenReductions().then(function (results) {
+      results.forEach(function (result) {
+        expect(['ACTIVE', 'SCHEDULED', null]).to.include(result.status)
+      })
+      done()
+    })
+  })
```
Comment (caoimheb) [inline test/integration/services/data/test-get-all-open-reductions.js]: Run a knex query to pull back all the open reductions and verify your results is the same set?

File: test/integration/services/data/test-update-reduction-status-by-ids.js [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+
+const appWorkloadOwnerHelper = require('../../../helpers/data/app-workload-owner-helper')
+const appReductionsHelper = require('../../../helpers/data/app-reductions-helper')
+
+const updateReductionStatusByIds = require('../../../../app/services/data/update-reduction-status-by-ids')
+const getAllOpenReductions = require('../../../../app/services/data/get-all-open-reductions')
+
+var inserts = []
+
+describe('services/data/get-all-open-reductions', function () {
```
Comment (caoimheb) [inline test/integration/services/data/test-update-reduction-status-by-ids.js]: Copy and paste? Also in the test it() below

File: app/services/data/get-contracted-hours.js [hunk 12]
```diff
@@ -3,7 +3,17 @@ const knex = require('knex')(knexConfig)
 
 module.exports = function (workloadOwnerId) {
   return knex('workload_owner').withSchema('app')
-    .where('id', workloadOwnerId)
-    .first('contracted_hours')
-    .then((result) => result.contracted_hours)
+    .crossJoin('workload_points')
+    .whereNull('workload_points.effective_to')
+    .andWhere('workload_owner.id', workloadOwnerId)
+    .first(['workload_owner.contracted_hours AS contracted_hours',
```
[the remark this answers is not in the document]
Comment (tsheppard01) [reply app/services/data/get-contracted-hours.js]: Don't think so I'll remove

File: app/services/data/get-current-workload-report.js (hunk 5 above)
Comment (tsheppard01) [reply app/services/data/get-current-workload-report.js]: renamed

File: app/services/data/get-current-workload-report.js (hunk 3 above)
Comment (tsheppard01) [reply app/services/data/get-current-workload-report.js]: added

File: app/services/data/get-current-workload-report.js (hunk 4 above)
Comment (tsheppard01) [reply app/services/data/get-current-workload-report.js]: removed

File: app/services/data/update-workload-report-with-status.js (hunk 6 above)
Comment (tsheppard01) [reply app/services/data/update-workload-report-with-status.js]: thanks

File: app/services/workers/reductions-worker.js (hunk 8 above)
Comment (tsheppard01) [reply app/services/workers/reductions-worker.js]: Added, also added unit test

File: app/services/workers/reductions-worker.js (hunk 7 above)
Comment (tsheppard01) [reply app/services/workers/reductions-worker.js]: fixed

File: app/services/workers/create-workload.js (hunk 9 above)
Comment (tsheppard01) [reply app/services/workers/create-workload.js]: added

File: test/integration/services/data/test-get-all-open-reductions.js (hunk 10 above)
Comment (tsheppard01) [reply test/integration/services/data/test-get-all-open-reductions.js]: Change to do something similar, get all reductions, then verify that the ones that the original query didn't bring back are what I expect

File: test/integration/services/data/test-update-reduction-status-by-ids.js (hunk 11 above)
Comment (tsheppard01) [reply test/integration/services/data/test-update-reduction-status-by-ids.js]: How embarassing, fixed

**Pull request merged**